### DISCUSSION TITLE
disabled property not updating in listbox headless component

### DISCUSSION THREAD

**Discussion by jfsullivan** [alpinejs/alpine#3442] [7. Bugs (archived)]
I am using the **Listbox Headless UI** component but the **disabled** property is not able to be changed after initialization.  I am trying to dynamically disable some options based upon values of other inputs but it does not work.

Here is the example component from the documentation with a button added.  Upon clicking the button, it should change the name and the disabled property for the Phoenix framework option to name:**updated** and disabled:**false**.

It updates the frameworks property as shown in the span tag below the button.  The option label is updated in the component but the disabled setting on the last option does not change.

```
<div
    x-data="{
        value: null,
        frameworks: [
            {
                id: 1,
                name: 'Laravel',
                disabled: false,
            },
            {
                id: 2,
                name: 'Ruby on Rails',
                disabled: false,
            },
            {
                id: 3,
                name: 'Phoenix',
                disabled: true,
            },
        ],
    }"
>
    <div
        x-listbox
        x-model="value"
        class="relative"
    >
        <label x-listbox:label class="sr-only">Backend framework</label>

        <button
            x-listbox:button
            class="flex items-center justify-between gap-2 w-52 bg-white pl-5 pr-3 py-2.5 rounded-md shadow"
        >
            <span x-text="value ? value.name : 'Select framework'" class="truncate"></span>

            <!-- Heroicons up/down -->
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="shrink-0 w-5 h-5 text-gray-500"><path fill-rule="evenodd" d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z" clip-rule="evenodd" /></svg>
        </button>

        <ul
            x-listbox:options
            x-transition.origin.top.right
            x-cloak
            class="absolute right-0 w-48 mt-2 z-10 origin-top-right bg-white border border-gray-200 divide-y divide-gray-100 rounded-md shadow-md outline-none"
        >
            <template x-for="framework in frameworks" :key="framework.id">
                <li
                    x-listbox:option
                    :value="framework"
                    :disabled="framework.disabled"
                    :class="{
                        'bg-cyan-500/10 text-gray-900': $listboxOption.isActive,
                        'text-gray-600': ! $listboxOption.isActive,
                        'opacity-50 cursor-not-allowed': $listboxOption.isDisabled,
                    }"
                    class="flex items-center cursor-default justify-between gap-2 w-full px-4 py-2 text-sm transition-colors"
                >
                    <span x-text="framework.name"></span>

                    <span x-show="$listboxOption.isSelected" class="text-cyan-600 font-bold">&check;</span>
                </li>
            </template>
        </ul>
    </div>
    <button x-on:click="frameworks[2].disabled = false; frameworks[2].name='Last Option New Name';">Enable Last Option</button>
    <span x-text="JSON.stringify(frameworks)"></span>
</div>
```

**Initial Settings**
![image](https://user-images.githubusercontent.com/5075028/222323135-ea46cd21-61a2-40bd-87eb-bd6a760faa5e.png)

**Updated Settings**
![image](https://user-images.githubusercontent.com/5075028/222323307-ed4210ec-173f-45ec-ba1b-0ea809a5c725.png)

Thank you for any help.

**Comment by ekwoka** [comment]
is it changing the disabled state on the item? and not updating the css?

checked and yes that's the case. idk what `$listboxoption` does for it's `isDisabled` but I don't see why it's even helpful?

There are other issues here as well. The `li` need to have `button` elements, then you can just find `disabled` how it is there, and have your styles to be attached to the `disabled` pseudo selector.

Based on what I can tell in the source, the `isDisabled` never updates. Why? I don't know. There are lots of issues with the implementation.

I'll see if I can give it some time to update it to solve this issue (And bring some other benefits along)

**Comment by jfsullivan** [comment]
The isDisabled is used to alter the css to make it looked disabled with this line....

`
'opacity-50 cursor-not-allowed': $listboxOption.isDisabled,
`

But it also prevents that option from being selected so there is some logic happening with it.  I can just use logic based upon framework.disabled to alter the css but the option would still be selectable.

Thank you for looking at it and offering to help.  My limited knowledge of javascript prevents me from fully understanding the component logic.

**Reply by ekwoka** [reply]
you don't need to use logic in JS at all to change the styling.

You should only use css to do that.

And they should be buttons, not LI, since you can't tab to `li` (or at least, put button in the li)
